Mt7921e devel - MTK internal SQC fixes

objelf opened the pull request.

moore-bros committed:
mt76: mt7921: add deep sleep control to runtime-pm knob

Add addtional the deep sleep control to runtime-pm knob to
allow us to control driver switching between the full power
mode and the deep sleep mode the firmware is able to support.

Signed-off-by: Sean Wang <sean.wang@mediatek.com>

Commit:
mt76: connac: fix WoW with disconnetion and bitmap pattern

Update MCU command usage to fix WoW configuration with disconnection
and bitmap pattern and to avoid magic number.

Fixes: ffa1bf97425 ("mt76: mt7921: introduce PM support")
Signed-off-by: YN Chen <yn.chen@mediatek.com>
Signed-off-by: Sean Wang <sean.wang@mediatek.com>

moore-bros committed:
mt76: connac: fw_own rely on all packet memory all being free

If the device is MMIO-based, we must ensure all TxD/TxP on the host
memory all being consumed by the device prior to safely switching to
fw_own state.

Fixes: ec7bd7b4a9c0 ("mt76: connac: check wake refcount in mcu_fw_pmctrl")
Signed-off-by: Sean Wang <sean.wang@mediatek.com>

moore-bros committed:
mt76: mt7921: fix mt7921_wfsys_reset sequence

WiFi subsytem reset should control MT_WFSYS_SW_RST_B and then poll the
same register until the bit WFSYS_SW_INIT_DONE bit is set.

Fixes: 0c1ce9884607 ("mt76: mt7921: add wifi reset support")
Signed-off-by: Sean Wang <sean.wang@mediatek.com>

moore-bros committed:
mt76: mt7921: enable deep sleep at runtime

Enable the deep sleep mode with that firmware is able to trap into
the doze state at runtime to reduce the power consumption further.

The deep sleep mode is not allowed in the STA state transition with
the firmware to have the fast connection experience as we've done in
the full power mode

Signed-off-by: Sean Wang <sean.wang@mediatek.com>

moore-bros committed:
mt76: mt7921: fix reset under the deep sleep is enabled

To fix possibly the race to access register between the WiFi reset
and the other context that is caused by explicitly cancelling ps_work
and wake_work to break PM_STATE consistency.

Deep sleep would cause the hardware into the inactive state,
so we forcely put device drv_own state before we start to reset.

The patch also ignore the reset request when the procedure is in
progress to avoid the consecutive WiFi resets.

localhost ~ # [ 2932.073966] SError Interrupt on CPU7, code 0xbe000011 -- SError
[ 2932.073967] CPU: 7 PID: 8761 Comm: kworker/u16:2 Not tainted 5.4.112 #30
[ 2932.073968] Hardware name: MediaTek Asurada rev1 board (DT)
[ 2932.073968] Workqueue: phy0 ieee80211_reconfig_filter [mac80211]
[ 2932.073969] pstate: 80400089 (Nzcv daIf +PAN -UAO)
[ 2932.073969] pc : el1_irq+0x78/0x180
[ 2932.073970] lr : mt76_mmio_rmw+0x30/0x5c [mt76]
[ 2932.073970] sp : ffffffc01142bad0
[ 2932.073970] x29: ffffffc01142bc00 x28: ffffff8f96fb1e00
[ 2932.073971] x27: ffffffd2cdc12138 x26: ffffffd2cdaeb018
[ 2932.073972] x25: 0000000000000000 x24: ffffff8fa8e14c08
[ 2932.073973] x23: 0000000080c00009 x22: ffffffd2a5603918
[ 2932.073974] x21: ffffffc01142bc10 x20: 0000007fffffffff
[ 2932.073975] x19: 0000000000000000 x18: 0000000000000400
[ 2932.073975] x17: 0000000000000400 x16: ffffffd2cd2b87dc
[ 2932.073976] x15: 0000000000000000 x14: 0000000000000000
[ 2932.073977] x13: 0000000000000001 x12: 0000000000000001
[ 2932.073978] x11: 0000000000000001 x10: 000000000010e000
[ 2932.073978] x9 : 0000000000000000 x8 : ffffffc013921404
[ 2932.073979] x7 : 000000b2b5593519 x6 : 0000000000300000
[ 2932.073980] x5 : 0000000000000000 x4 : ffffffc01142bbc8
[ 2932.073980] x3 : 00000000000001f0 x2 : 0000000000000000
[ 2932.073981] x1 : 0000000000021404 x0 : ffffff8fa8e12300
[ 2932.073982] Kernel panic - not syncing: Asynchronous SError Interrupt
[ 2932.073983] CPU: 7 PID: 8761 Comm: kworker/u16:2 Not tainted 5.4.112 #30
[ 2932.073983] Hardware name: MediaTek Asurada rev1 board (DT)
[ 2932.073984] Workqueue: phy0 ieee80211_reconfig_filter [mac80211]
[ 2932.073984] Call trace:
[ 2932.073985]  dump_backtrace+0x0/0x14c
[ 2932.073985]  show_stack+0x20/0x2c
[ 2932.073985]  dump_stack+0xa0/0xf8
[ 2932.073986]  panic+0x154/0x360
[ 2932.073986]  test_taint+0x0/0x44
[ 2932.073986]  arm64_serror_panic+0x78/0x84
[ 2932.073987]  do_serror+0x0/0x118
[ 2932.073987]  do_serror+0xa4/0x118
[ 2932.073987]  el1_error+0x84/0xf8
[ 2932.073988]  el1_irq+0x78/0x180
[ 2932.073988]  mt76_mmio_rr+0x30/0xf0 [mt76]
[ 2932.073988]  mt76_mmio_rmw+0x30/0x5c [mt76]
[ 2932.073989]  mt7921_rmw+0x4c/0x5c [mt7921e]
[ 2932.073989]  mt7921_configure_filter+0x138/0x160 [mt7921e]
[ 2932.073990]  ieee80211_configure_filter+0x2f0/0x3e0 [mac80211]
[ 2932.073990]  ieee80211_reconfig_filter+0x1c/0x28 [mac80211]
[ 2932.073990]  process_one_work+0x208/0x3c8
[ 2932.073991]  worker_thread+0x23c/0x3e8
[ 2932.073991]  kthread+0x140/0x17c
[ 2932.073992]  ret_from_fork+0x10/0x18
[ 2932.074071] SMP: stopping secondary CPUs
[ 2932.074071] Kernel Offset: 0x12bc800000 from 0xffffffc010000000
[ 2932.074072] PHYS_OFFSET: 0xfffffff180000000
[ 2932.074072] CPU features: 0x080026,2a80aa18
[ 2932.074072] Memory Limit: none

Fixes: 87843566e581 ("mt76: mt7921: enable deep sleep at runtime")
Signed-off-by: Sean Wang <sean.wang@mediatek.com>

moore-bros committed:
mt76: add MT76_STATE_WOW to indicate the completion of the WoW setup

The patch just replaces MT76_STATE_SUSPEND with MT76_STATE_WOW and
don't have any logic change.

This is a preliminary patch for we want to reserve the MT76_STATE_SUSPEND
to indicate the driver in suspend to get rid of the worker still possibly
accessing the bus.

Signed-off-by: Sean Wang <sean.wang@mediatek.com>

moore-bros committed:
mt76: mt7921: fix invalid register access in wake_work

Make sure mt7921_pm_wake_work wouldn't be scheduled after the driver is
in suspend mode to fix the following the kernel crash.

[ 3515.390012] mt7921e 0000:01:00.0: calling pci_pm_suspend+0x0/0x22c @ 2869, parent: 0000:00:00.0
[ 3515.390015] mt7921e 0000:01:00.0: mt7921_pci_suspend +
[ 3515.396395] anx7625 3-0058: anx7625_suspend+0x0/0x6c returned 0 after 0 usecs
[ 3515.405965] mt7921e 0000:01:00.0: mt7921_pci_suspend -
[ 3515.411336] usb 1-1.4: usb_dev_suspend+0x0/0x2c returned 0 after 1 usecs
[ 3515.411513] SError Interrupt on CPU7, code 0xbe000011 -- SError
[ 3515.411515] CPU: 7 PID: 2849 Comm: kworker/u16:27 Not tainted 5.4.114 #44
[ 3515.411516] Hardware name: MediaTek Asurada rev1 board (DT)
[ 3515.411517] Workqueue: mt76 mt7921_pm_wake_work [mt7921e]
[ 3515.411518] pstate: 80c00009 (Nzcv daif +PAN +UAO)
[ 3515.411519] pc : mt76_mmio_rr+0x30/0xf0 [mt76]
[ 3515.411520] lr : mt7921_rr+0x38/0x44 [mt7921e]
[ 3515.411520] sp : ffffffc015813c50
[ 3515.411521] x29: ffffffc015813c50 x28: 0000000000000402
[ 3515.411522] x27: ffffffe5a2012138 x26: ffffffe5a1eea018
[ 3515.411524] x25: 00000000328be505 x24: 00000000000a0002
[ 3515.411525] x23: 0000000000000006 x22: ffffffbd29b7a300
[ 3515.411527] x21: ffffffbd29b7a300 x20: 00000000000e0010
[ 3515.411528] x19: 00000000eac08f43 x18: 0000000000000000
[ 3515.411529] x17: 0000000000000000 x16: ffffffe5a16b2914
[ 3515.411531] x15: 0000000000000010 x14: 0000000000000010
[ 3515.411532] x13: 00000000003dd3a2 x12: 0000000000010000
[ 3515.411533] x11: ffffffe597abec14 x10: 0000000000000010
[ 3515.411535] x9 : ffffffe597abeba8 x8 : ffffffc013ce0010
[ 3515.411536] x7 : 000000b2b5593519 x6 : 0000000000300000
[ 3515.411537] x5 : 0000000000000000 x4 : 0000000000000032
[ 3515.411539] x3 : 0000000000000000 x2 : 0000000000000004
[ 3515.411540] x1 : 00000000000e0010 x0 : ffffffbd29b7a300
[ 3515.411542] Kernel panic - not syncing: Asynchronous SError Interrupt
[ 3515.411543] CPU: 7 PID: 2849 Comm: kworker/u16:27 Not tainted 5.4.114 #44
[ 3515.411544] Hardware name: MediaTek Asurada rev1 board (DT)
[ 3515.411544] Workqueue: mt76 mt7921_pm_wake_work [mt7921e]
[ 3515.411545] Call trace:
[ 3515.411546]  dump_backtrace+0x0/0x14c
[ 3515.411546]  show_stack+0x20/0x2c
[ 3515.411547]  dump_stack+0xa0/0xfc
[ 3515.411548]  panic+0x154/0x350
[ 3515.411548]  panic+0x0/0x350
[ 3515.411549]  arm64_serror_panic+0x78/0x84
[ 3515.411550]  do_serror+0x0/0x118
[ 3515.411550]  do_serror+0xa4/0x118
[ 3515.411551]  el1_error+0x84/0xf8
[ 3515.411552]  mt76_mmio_rr+0x30/0xf0 [mt76]
[ 3515.411552]  mt7921_rr+0x38/0x44 [mt7921e]
[ 3515.411553]  __mt76_poll_msec+0x5c/0x9c [mt76]
[ 3515.411554]  __mt7921_mcu_drv_pmctrl+0x50/0x94 [mt7921e]
[ 3515.411555]  mt7921_mcu_drv_pmctrl+0x38/0xb0 [mt7921e]
[ 3515.411555]  mt7921_pm_wake_work+0x34/0xd4 [mt7921e]
[ 3515.411556]  process_one_work+0x208/0x3c8
[ 3515.411557]  worker_thread+0x23c/0x3e8
[ 3515.411557]  kthread+0x144/0x178
[ 3515.411558]  ret_from_fork+0x10/0x18
[ 3515.418831] SMP: stopping secondary CPUs
[ 3515.418832] Kernel Offset: 0x2590c00000 from 0xffffffc010000000
[ 3515.418832] PHYS_OFFSET: 0xffffffc400000000
[ 3515.418833] CPU features: 0x080026,2a80aa18
[ 3515.418834] Memory Limit: none
[DL] 00000000 00000000 010701

Signed-off-by: Sean Wang <sean.wang@mediatek.com>

moore-bros committed:
mt76: mt7921: fix UC entry is being overwritten

Fix UC entry is being overwritten by BC entry

Fixes: 36fcc8cff592 ("mt76: mt7921: introduce mt7921_mcu_sta_add routine")
Signed-off-by: Deren Wu <deren.wu@mediatek.com>
Signed-off-by: YN Chen <yn.chen@mediatek.com>
Signed-off-by: Sean Wang <sean.wang@mediatek.com>

moore-bros committed:
mt76: mt7921: consider the invalid value for to_rssi

It is possible the RCPI from the certain antenna is an invalid value,
especially packets are receiving while the system is frequently entering
deep sleep mode, so consider calculating RSSI with the reasonable upper
bound to avoid report the wrong value to the mac80211 layer.

Fixes: 163f4d22c118 ("mt76: mt7921: add MAC support")
Signed-off-by: Sean Wang <sean.wang@mediatek.com>

moore-bros committed:
mt76: mt7921: add back connection monitor support

Hw beacon cmd to the mt7921 firmware doesn't only filter out the beacon,
but also performs its own connection monitoring, including periodic
keep-alives to the AP and probing the AP on beacon loss. Will indicate
the host with the event when the firmware detects the connection is lost.

Fixes: 1d8efc741df8 ("mt76: mt7921: introduce Runtime PM support")
Signed-off-by: Deren Wu <deren.wu@mediatek.com>
Signed-off-by: YN Chen <yn.chen@mediatek.com>
Signed-off-by: Sean Wang <sean.wang@mediatek.com>

LorenzoBianconi commented on drivers/net/wireless/mediatek/mt76/mt7921/regs.h:
```diff
@@ -36,18 +36,6 @@
-#define MT_MDP_RCFR1_RX_DROPPED_MCAST	GENMASK(30, 29)
-#define MT_MDP_TO_HIF			0
-#define MT_MDP_TO_WM			1
-
```
@objelf I would not drop these definitions even if they are not used at the moment, maybe we need them in the future

objelf replied:
ok, lets keep it.

LorenzoBianconi commented on drivers/net/wireless/mediatek/mt76/mt76_connac.h:
```diff
@@ -134,12 +134,16 @@ mt76_connac_pm_unref(struct mt76_connac_pm *pm)
 
+	if (mt76_is_mmio(dev) && dev->token_count)
+		return true;
+
```
@objelf I guess you need to check token_count before setting MT76_STATE_PM otherwise we will return true (so not performing fw_own) and state is marked in PM state. Another approach would be to clear MT76_STATE_PM before returning true

objelf replied:
ack i will check about it.

LorenzoBianconi commented on drivers/net/wireless/mediatek/mt76/mt7921/mcu.c:
```diff
@@ -981,6 +981,7 @@ int mt7921_run_firmware(struct mt7921_dev *dev)
 	set_bit(MT76_STATE_MCU_RUNNING, &dev->mphy.state);
 	mt7921_mcu_fw_log_2_host(dev, 1);
+	mt76_connac_mcu_set_deep_sleep(&dev->mt76, dev->pm.enable);
 
```
@objelf this is not correct since after mt7921_mcu_init() there are other device accesses without grabbing dev mutex, I guess we need to move it at the end of mt7921_register_device

objelf replied:
ack, i will check about it.

LorenzoBianconi commented on drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.h:
```diff
@@ -1090,5 +1090,7 @@ int mt76_connac_mcu_set_roc(struct mt76_dev *dev, struct ieee80211_vif *vif,
 			       struct mt76_connac_roc *roc,
 			       struct mt76_connac_roc_tlv *event);
-
+int mt76_connac_sta_state_dp(struct mt76_dev *dev,
```
@objelf I think this will not apply upstream :)

objelf replied:
do you mean it is not neccessary to remove the blank line, right ?

LorenzoBianconi replied:
@objelf I mean mt76_connac_mcu_roc_event is not upstream yet, but no worries I will rebase it, just send me an updated PR on top of mt7921e-devel branch

objelf replied:
ok, thx

LorenzoBianconi commented on drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c:
```diff
@@ -1552,6 +1552,21 @@ int mt76_connac_mcu_set_deep_sleep(struct mt76_dev *dev, bool enable)
+}
+EXPORT_SYMBOL_GPL(mt76_connac_sta_state_dp);
+
 void mt76_connac_mcu_coredump_event(struct mt76_dev *dev, struct sk_buff *skb,
```
@objelf in order to reduce unnecessary mcu cmds we can do something like:
enable deep sleep switching from IEEE80211_STA_NONE to IEEE80211_STA_NOTEXIST and from IEEE80211_STA_ASSOC to IEEE80211_STA_AUTHORIZED while we can disable it switching from IEEE80211_STA_NOTEXIST to IEEE80211_STA_NONE, what do you think?

objelf replied:
ok, i will try this way and do some stress test.

LorenzoBianconi commented on drivers/net/wireless/mediatek/mt76/mt7921/debugfs.c:
```diff
@@ -236,6 +236,19 @@ mt7921_pm_set(void *data, u64 val)
+		mt7921_mutex_release(dev);
+		return 0;
+	}
+
```
@objelf I would prefer a dedicated debugs knob for enabling/disabling deep sleep avoid magic numbers

objelf replied:
ack. i will create dedicated debugs knob for enabling and disabling deep sleep.

LorenzoBianconi commented on drivers/net/wireless/mediatek/mt76/mt7921/mac.c:
```diff
@@ -1275,29 +1276,34 @@ mt7921_mac_reset(struct mt7921_dev *dev)
 			break;
 	}
 	mutex_unlock(&dev->mt76.mutex);
-
```
@objelf unnecessary change

LorenzoBianconi commented on drivers/net/wireless/mediatek/mt76/mt7921/mac.c:
```diff
@@ -1275,29 +1276,34 @@ mt7921_mac_reset(struct mt7921_dev *dev)
 {
-	struct ieee80211_hw *hw;
-	struct mt7921_dev *dev;
+	struct mt7921_dev *dev = container_of(work, struct mt7921_dev, reset_work);
```
@objelf please check # of columns here

LorenzoBianconi commented on drivers/net/wireless/mediatek/mt76/mt7921/mac.c:
```diff
@@ -1275,29 +1276,34 @@ mt7921_mac_reset(struct mt7921_dev *dev)
+	cancel_delayed_work_sync(&pm->ps_work);
+	cancel_work_sync(&pm->wake_work);
+	mt7921_mcu_drv_pmctrl(dev);
+
```
@objelf is it enough running __mt7921_mcu_drv_pmctrl() below?

LorenzoBianconi commented on drivers/net/wireless/mediatek/mt76/mt76_connac_mac.c:
```diff
@@ -13,6 +13,9 @@ int mt76_connac_pm_wake(struct mt76_phy *phy, struct mt76_connac_pm *pm)
 
+	if (test_bit(MT76_STATE_RESET, &phy->state))
+		return 0;
+
```
@objelf I guess this approach is a bit racy because we can run drv_own or fw_own while running the reset right?

objelf replied:
drv_own and fw_own don't work during wifisys reset

LorenzoBianconi replied:
in the code flowpath we can have both mt7921_reset and fw_own/drv_own running at the same time, right (e.g. on different CPUs). e.g what happen if mt76_connac_pm_wake has already scheduled the wake work?

LorenzoBianconi replied:
@objelf sorry this comment is related to the change "https://github.com/LorenzoBianconi/wireless-drivers-next/pull/33/commits/0712b8987b3f89947954f9a8819fe890b785c453". Please ignore it

LorenzoBianconi commented on drivers/net/wireless/mediatek/mt76/mt76.h:
```diff
@@ -313,6 +313,7 @@ enum {
 	MT76_READING_STATS,
 	MT76_STATE_POWER_OFF,
 	MT76_STATE_SUSPEND,
+	MT76_STATE_RESET,
```
@objelf we already have MT76_RESET

LorenzoBianconi commented on drivers/net/wireless/mediatek/mt76/mt7921/mac.c:
```diff
@@ -448,16 +448,19 @@ int mt7921_mac_fill_rx(struct mt7921_dev *dev, struct sk_buff *skb)
+		status->signal = -128;
+		for (i = 0; i < hweight8(mphy->antenna_mask); i++) {
+			if (!(status->chains & BIT(i)) ||
+			     (status->chain_signal[i] >= 0))
```
@objelf nitpick: no need for () around status->chain_signal[i] >= 0

LorenzoBianconi commented on drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c:
```diff
@@ -837,10 +837,13 @@ int mt76_connac_mcu_add_sta_cmd(struct mt76_phy *phy,
+	if (info->sta) {
+		mt76_connac_mcu_sta_basic_tlv(skb, info->vif, info->sta,
+					      info->enable);
+		if (info->enable && info->sta)
```
@objelf we do not need to check info->sta here

objelf replied:
ok, wil remove the redundant check.

LorenzoBianconi commented on drivers/net/wireless/mediatek/mt76/mt76_connac_mcu.c:
```diff
@@ -837,10 +837,13 @@ int mt76_connac_mcu_add_sta_cmd(struct mt76_phy *phy,
 	if (IS_ERR(skb))
 		return PTR_ERR(skb);
 
-	mt76_connac_mcu_sta_basic_tlv(skb, info->vif, info->sta, info->enable);
```
@objelf this is even used in __mt7615_mcu_add_sta(). Can you please check if this change is fine for 7663?

objelf replied:
Deren mentioned the cmd  usage of mt7921 should be aligned to mt7663

moore-bros committed:
mt76: mt7921: Don't alter Rx path classifier

Keep Rx path classifier the mt7921 firmware prefers to allow frames pass
through MCU.

Fixes: 5c14a5f944b9 ("mt76: mt7921: introduce mt7921e support")
Signed-off-by: Sean Wang <sean.wang@mediatek.com>

Commit:
mt76: connac: add bss color support for sta mode

Add bss color support for sta mode

Signed-off-by: Jayden.Kuo <jayden.kuo@mediatek.com>
Signed-off-by: YN Chen <yn.chen@mediatek.com>
Signed-off-by: Sean Wang <sean.wang@mediatek.com>

objelf commented:
rebase and update with v2
